### PR TITLE
More dropdown tests

### PR DESCRIPTION
**Pull request by mcbouslog**
Almost completes #3298.

Adds additional tests for Dropdown Task, including tests for multiple, conditional selects.

# Review Checklist

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [x] Can you `rm -rf node_modules/ && npm install` and app works as expected?
- [x] Did you deploy a staging branch? - https://more-dropdown-tests.pfe-preview.zooniverse.org/
- [ ] Did you get any type checking errors from [Babel Typecheck](https://github.com/codemix/babel-plugin-typecheck)

## Optional

- [ ] If it's a new component, is it in ES6? Is it clear of warnings from ESLint?
- [ ] Have you replaced any `ChangeListener` or `PromiseRenderer` components with code that updates component state?
- [ ] If changes are made to the classifier, does the dev classifier still work?
- [ ] Have you added in [flow type annotations](https://flowtype.org/docs/type-annotations.html)?

### CODE REVIEW

**Comment by mcbouslog** [comment]
This is the largest amount of testing I've done to date, please provide any feedback that comes to mind, there's nothing too small, and I wouldn't be surprised if a certain section or sections could use entire rewriting.

The edge case related to the commented out (failing) asserts is for when there's a select that is conditional to a select that is not directly proceeding it (like in Dev Classifier - Best State Team select and State select). If the selects between the noted selects are answered, it changes the `optionsKeys` in a way that prevents the proper options for the conditional select (Best State Team) to load. In the Dev Classifier select a country and state, then note the options for Best State Team, then select a County and City, now note the options for Best State Team are gone. Or make selections in all selects, including a custom answer or not in Best State Team, click Next, then go Back, and you'll see the stored annotations aren't loading correctly.

I have a solution for the edge case (which I don't think any project is currently experiencing as it's an obscure case) that I think I can implement within the next day or two.

File: app/classifier/tasks/dropdown/index.cjsx [hunk 1]
```diff
@@ -160,7 +161,7 @@ module.exports = React.createClass
           disabled = @getDisabledAttribute(i)
           selectedOption = if @state.selectedOptions[i]?.value then @state.selectedOptions[i] else null
           optionsKeys.push selectedOption.value if selectedOption?.value
-          <div key={select.id}>
+          <div id={select.id} key={select.id}>
```
Comment (mcbouslog) [inline app/classifier/tasks/dropdown/index.cjsx]: helps with testing `wrapper.find('#countryID')`

File: app/classifier/tasks/dropdown/index.spec.js [hunk 2]
```diff
@@ -1,66 +1,263 @@
...
+
 import React from 'react';
 import assert from 'assert';
 import { mount } from 'enzyme';
+import Select from 'react-select'; // required to properly simulate change
 import DropdownTask from './';
+import { workflow } from '../../../pages/dev-classifier/mock-data';
 
-const task = {
+const singleSelect = {
```
Comment (mcbouslog) [inline app/classifier/tasks/dropdown/index.spec.js]: should just use multiSelect? not sure it could encompass all scenarios. I like having `singleSelect`, but not if there's too much going on.

File: app/classifier/tasks/dropdown/index.spec.js [hunk 3]
```diff
@@ -1,66 +1,263 @@
...
-  value: [{option: true, value: 3}]
-};
+const multiSelects = workflow.tasks.dropdown;
 
-describe('DropdownTask', function () {
-  let wrapper;
+// multiSelects:
+//   1 - Country (required:true)
+//   2 - State (condition:Country, required:true, allowCreate:false)
+//   3 - County (condition:State, allowCreate:true)
```
Comment (mcbouslog) [inline app/classifier/tasks/dropdown/index.spec.js]: rename all instances of `county` (in mock data as well) to `region` to make things more clear, less similar to `country`?

File: app/classifier/tasks/dropdown/index.spec.js [hunk 4]
```diff
@@ -1,66 +1,263 @@
...
+        wrapper = mount(<DropdownTask task={multiSelects} annotation={annotation} onChange={function (a) { return a; }} />);
+      });
+
+      it('should make selects conditional on first annotation and allowCreate false now enabled (State now enabled, City still disabled)', function () {
+        assert.equal(wrapper.find('.is-disabled').length, 1);
+      });
+
+      it('should show the proper options in selects conditional (State)', function () {
+        const stateSelect = wrapper.find('#stateID').find(Select);
+        // TODO is deepEqual ok to use here? alternative could be to check length, individual objects' contents
```
Comment (mcbouslog) [inline app/classifier/tasks/dropdown/index.spec.js]: not sure I'm properly using `deepEqual` vs `equal`

File: app/classifier/tasks/dropdown/index.spec.js [hunk 5]
```diff
@@ -1,66 +1,263 @@
...
+      it('should clear related selects (County, City, Team) when conditional select (State) changed', function () {
+        wrapper.instance().onChangeSelect(1, multiSelects.selects[1].options['Mypos-value'][2]);
+        const expectedAnnotation = { value: [
+          { value: 'Mypos-value', option: true },
+          { value: 'Shire-value', option: true },
+          { value: null, option: false },
+          { value: null, option: false },
+          { value: null, option: false }
+        ] };
+        // TODO is deepEqual ok to use here? alternative could be to check length, individual objects' contents
```
Comment (mcbouslog) [inline app/classifier/tasks/dropdown/index.spec.js]: not sure I'm properly using `deepEqual` vs `equal`

File: app/classifier/tasks/dropdown/index.spec.js [hunk 6]
```diff
@@ -1,66 +1,263 @@
...
+      it('should clear related selects (County, City, Team) when conditional select (State) changed', function () {
+        wrapper.instance().onChangeSelect(1, multiSelects.selects[1].options['Canada-value'][1]);
+        const expectedAnnotation = { value: [
+          { value: 'Canada-value', option: true },
+          { value: 'ON', option: true },
+          { value: null, option: false },
+          { value: null, option: false },
+          { value: null, option: false }
+        ] };
+        // TODO is deepEqual ok to use here? alternative could be to check length, individual objects' contents
```
Comment (mcbouslog) [inline app/classifier/tasks/dropdown/index.spec.js]: not sure I'm properly using `deepEqual` vs `equal`

**Comment by mcbouslog** [comment]
edits made to `mock-data` Dropdown Task for increased clarity

File: app/classifier/tasks/dropdown/index.cjsx (hunk 1 above)
Comment (eatyourgreens) [inline app/classifier/tasks/dropdown/index.cjsx]: I think you can find a component by key. Something like `wrapper.find(div).find({key: select.id})`

File: app/classifier/tasks/dropdown/index.spec.js [hunk 7]
```diff
@@ -1,66 +1,263 @@
...
+//   5 - Best State Team (condition:State, allowCreate:true)
+
+describe('DropdownTask:static methods', function () {
+  it('should have the correct question text', function () {
+    assert.equal(DropdownTask.getTaskText(singleSelect), singleSelect.instruction);
+  });
 
-  beforeEach(function () {
-    wrapper = mount(<DropdownTask task={task} annotation={annotation} onChange={function(a) {return a;} } />);
+  it('the default annotation should be an empty array', function () {
```
Comment (eatyourgreens) [inline app/classifier/tasks/dropdown/index.spec.js]: Small note on style: "it the default annotation should be…" doesn't read very well so maybe reword this test?

File: app/classifier/tasks/dropdown/index.spec.js [hunk 8]
```diff
@@ -1,66 +1,263 @@
...
+          { value: null, option: false },
+          { value: null, option: false },
+          { value: null, option: false }
+        ] };
+
+        wrapper = mount(<DropdownTask task={multiSelects} annotation={annotation} onChange={function (a) { return a; }} />);
+      });
+
+      it('should make selects conditional on first annotation and allowCreate false now enabled (State now enabled, City still disabled)', function () {
+        assert.equal(wrapper.find('.is-disabled').length, 1);
```
Comment (eatyourgreens) [inline app/classifier/tasks/dropdown/index.spec.js]: This will also pass if City is enabled and State is disabled.

File: app/classifier/tasks/dropdown/index.cjsx [hunk 9]
```diff
@@ -92,6 +92,7 @@ module.exports = React.createClass
     annotationValues = @props.annotation.value
     unless annotationValues.length
       @props.annotation.value = @props.task.selects.map -> {value: null, option: false}
+      @syncAnnotations()
```
Comment (eatyourgreens) [inline app/classifier/tasks/dropdown/index.cjsx]: Is this fixing a bug?

**Comment by eatyourgreens** [comment]
Re. the failing test, maybe write that up as an issue, and leave the test (and fix) for it to a separate PR.

**Comment by eatyourgreens** [comment]
Dropdown bug written up in #3488

File: app/classifier/tasks/dropdown/index.cjsx (hunk 1 above)
Comment (mcbouslog) [inline app/classifier/tasks/dropdown/index.cjsx]: I thought so as well, but can't get it to work.
`wrapper.find('div').find('[id="countyID"]')` works, but
`wrapper.find('div').find('[key="countyID"]')` doesn't work.

Comment (eatyourgreens) [inline app/classifier/tasks/dropdown/index.cjsx]: `.find({key: 'countyID'})` should work. You can't use CSS selectors for React props. The g[uide to Enzyme Selectors](http://airbnb.io/enzyme/docs/api/selector.html) is really useful here.

Comment (mcbouslog) [inline app/classifier/tasks/dropdown/index.cjsx]: =/ Still can't get to work, not sure what I'm doing wrong. I'm sure something stupid right in front of me.

File: app/classifier/tasks/dropdown/index.cjsx (hunk 9 above)
Comment (mcbouslog) [inline app/classifier/tasks/dropdown/index.cjsx]: It is, but I like your idea of separate PRs, I'll move this bug fix to Issue #3488 as well, or can open a different issue.

**Comment by mcbouslog** [comment]
I removed failing tests and the `@syncAnnotations` edit to the actual task `index.cjsx` as I like your idea of separate PRs, this one (other than adding `id` to a div) doesn't change the actual dropdown task and just adds passing tests, so the next PR can focus on fixes and new tests.
